src/github.com/matrix-org/dendrite/roomserver/input/state.go

```diff
@@ -114,8 +116,94 @@ func calculateAndStoreStateAfterManyEvents(db RoomEventDatabase, roomNID types.R
 	return db.AddState(roomNID, nil, state)
 }
 
+// loadStateEvents loads the matrix events for a list of state entries.
+// Returns a list of events and a map from string event ID back to state entry.
+// The map can be used to recover which numeric state entry a given event is for.
+// Returns an error if there was a problem talking to the database.
+func loadStateEvents(db RoomEventDatabase, entries []types.StateEntry) ([]gomatrixserverlib.Event, map[string]types.StateEntry, error) {
+	eventNIDs := make([]types.EventNID, len(entries))
+	for i := range entries {
+		eventNIDs[i] = entries[i].EventNID
+	}
+	events, err := db.Events(eventNIDs)
+	if err != nil {
+		return nil, nil, err
+	}
+	eventIDMap := map[string]types.StateEntry{}
+	result := make([]gomatrixserverlib.Event, len(entries))
+	for i := range entries {
+		event, ok := eventMap(events).lookup(entries[i].EventNID)
+		if !ok {
+			panic(fmt.Errorf("Corrupt DB: Missing event numeric ID %d", entries[i].EventNID))
+		}
+		result[i] = event.Event
+		eventIDMap[event.Event.EventID()] = entries[i]
+	}
+	return result, eventIDMap, nil
+}
+
+// resolveConflicts resolves a list of conflicted state entries. It takes two lists.
+// The first is a list of all state entries both conflicted and not conflicted.
+// The second is a list of all state entries where there is more than one numeric event ID for the same state key tuple.
+// Returns a list of combining the entries without conflictes with the result of state resolution for the entries with conflicts.
```

Comment:
Also, *why* do we need to provide `combined` in the first place? It looks like we just get the non-conflicting ones and pass them through unaltered, so uhhh.. why can't the caller do that? The API shape feels weird currently.